Integer overflow in opj_t1_encode_cblks (src/lib/openjp2/t1.c)

(This problem is discovered when UBSAN is enabled)

on openjpeg 2.3 (latest version):
there is a integer overflow in the opj_t1_encode_cblks function (src/lib/openjp2/t1.c), which could be triggered by the POC.



POC address: https://github.com/ProbeFuzzer/poc/blob/master/openjpeg/openjpeg_2-3_opj_compress_integer-overflow_opj_t1_encode_cblks.bmp

Command: opj_compress -n 1 -i $POC -o /tmp/null.j2k
openjpeg/master/src/src/lib/openjp2/t1.c:2173:55: runtime error: signed integer overflow: 322385710 * 64 cannot be represented in type 'int'

```
/usr/local/bin/opj_compress -n 1 -i openjpeg_2-3_opj_compress_integer-overflow_opj_t1_encode_cblks.bmp -o /tmp/null.j2k

[INFO] tile number 1 / 1
[ERROR] opj_t2_encode_packet(): only 22 bytes remaining in output buffer. 359 needed.
[ERROR] Cannot encode tile
failed to encode image: opj_encode
failed to encode image: opj_end_compress
failed to encode image
```

So is it still affected?

@ProbeFuzzer, @pesc ,
the BMP image has bugs. I tried several programs:
//----
IMAGE HAS: num_colors (INT)-336860181 (UINT)3958107115

//----
opj_compress -i  openjpeg_2-3_opj_compress_integer-overflow_opj_t1_encode_cblks.bmp -o out.jp2

[ERROR] Number of resolutions is too high in comparison to the size of tiles
failed to encode image: opj_start_compress
failed to encode image: opj_encode
failed to encode image: opj_end_compress
failed to encode image

//----
gm animate openjpeg_2-3_opj_compress_integer-overflow_opj_t1_encode_cblks.bmp
(accepted)

//----
animate openjpeg_2-3_opj_compress_integer-overflow_opj_t1_encode_cblks.bmp
animate: insufficient image data in file `openjpeg_2-3_opj_compress_integer-overflow_opj_t1_encode_cblks.bmp' @ error/bmp.c/ReadBMPImage/677.

//----
bmptoppm -verbose openjpeg_2-3_opj_compress_integer-overflow_opj_t1_encode_cblks.bmp > bmp-to-ppm.txt
bmptoppm: Read 124 bytes of header
bmptoppm: BMP image header says:
bmptoppm:   Class of BMP: Windows (v5)
bmptoppm:   Width: 233 pixels
bmptoppm:   Height: 3 pixels
bmptoppm:   Depth: 1 planes
bmptoppm:   Row order: bottom up
bmptoppm:   Byte offset of raster within file: 138
bmptoppm:   Bits per pixel in raster: 32
bmptoppm:   Compression: none (bitfields)
bmptoppm:   Colors in color map: 0
bmptoppm: warning: some image data remains unread.
bmptoppm: WRITING PPM IMAGE
(accepted)

//----
I used 'openjpeg2-2018-12-26'.
winfried

--- a/src/lib/openjp2/t1.c
+++ b/src/lib/openjp2/t1.c
@@ -2168,9 +2168,18 @@ OPJ_BOOL opj_t1_encode_cblks(opj_t1_t *t1,
                         t1->data = tiledp;
                         t1->data_stride = tile_w;
                         if (tccp->qmfbid == 1) {
+                            /* Do multiplication on unsigned type, even if the
+                             * underlying type is signed, to avoid potential
+                             * int overflow on large value (the output will be
+                             * incorrect in such situation, but whatever...)
+                             * This assumes complement-to-2 signed integer
+                             * representation
+                             * Fixes https://github.com/uclouvain/openjpeg/issues/1053
+                             */
+                            OPJ_UINT32* OPJ_RESTRICT tiledp_u = (OPJ_UINT32*) tiledp;
                             for (j = 0; j < cblk_h; ++j) {
                                 for (i = 0; i < cblk_w; ++i) {
-                                    tiledp[tileIndex] *= (1 << T1_NMSEDEC_FRACBITS);
+                                    tiledp_u[tileIndex] <<= T1_NMSEDEC_FRACBITS;
                                     tileIndex++;
                                 }
                                 tileIndex += tileLineAdvance;

